File: script/tournaments.py
# originally made with chatgpt
# updates the two leaderboards for the weekly tournaments (trophies and most fish per week)

import re
from collections import defaultdict
import csv
from itertools import groupby

# Function to read renamed chatters from CSV file
def renamed(filename):
    renamed_chatters = {}
    with open('lists/renamed.csv', 'r') as file:
        reader = csv.DictReader(file)
        for row in reader:
            old_player = row['old_name']
            new_player = row['new_name']
            renamed_chatters[old_player] = new_player
    return renamed_chatters

# Function to read cheaters from a text file
def read_cheaters(filename):
    cheaters = []
    with open('lists/cheaters.txt', 'r') as file:
        # Read each line from the text file
        for line in file:
            # Append the line to the list of cheaters
            cheaters.append(line.strip())  # Strip any leading or trailing whitespace
    return cheaters

# Function to read verified players from a text file
def read_verified_players(filename):
    verified_players = []
    with open('lists/verified.txt', 'r') as file:
        # Read each line from the text file
        for line in file:
            # Append the line to the list of verified players
            verified_players.append(line.strip())  # Strip any leading or trailing whitespace
    return verified_players

renamed_chatters = renamed('lists/renamed.csv')
cheaters = read_cheaters('lists/cheaters.txt')
verified_players = read_verified_players('lists/verified.txt')

# Define a dictionary to store the maximum fish caught in a week for each player and the bot name
max_fish_in_week = defaultdict(lambda: {'fish_count': 0, 'bot': None})

# Define a dictionary to store the counts for each player's trophies and medals
player_counts = defaultdict(lambda: {'Trophy': 0, 'Silver': 0, 'Bronze': 0})

# Define point values
point_values = {'Trophy': 3, 'Silver': 1, 'Bronze': 0.5}

# Open and read the logs file
with open('logs/logs.txt', 'r', encoding='utf-8') as file:
    # Read each line in the file
    for line in file:

        # Find the player
        player_match = re.search(r'[@👥]\s?(\w+)', line)
        if player_match:
            player = player_match.group(1)
            
            # Skip processing for ignored players
            if player in cheaters:
                continue
            
            # Change to the latest name
            player = renamed_chatters.get(player, player)
            while player in renamed_chatters:
                player = renamed_chatters[player]

            # Get the amount of fish the player caught
            fish_match = re.search(r'(\d+) fish: (\w+)', line)
            if fish_match:
                fish_count = int(fish_match.group(1))
                bot_match = re.search(r'#\w+ \s?(\w+):', line)
                if bot_match:
                    bot = bot_match.group(1)
            
                # Update the record if the current fish count is greater
                if fish_count > max_fish_in_week[player]['fish_count']:
                    max_fish_in_week[player] = {'fish_count': fish_count, 'bot': bot}

            # Find all their medals and trophies
            achievements = re.findall(r'(Victory|champion|runner-up|third)', line)
            
            for achievement in achievements:
                if 'Victory' in achievement:
                    player_counts[player]['Trophy'] += 1
                elif 'runner-up' in achievement:
                    player_counts[player]['Silver'] += 1
                elif 'third' in achievement:
                    player_counts[player]['Bronze'] += 1
                elif 'champion' in achievement:
                    player_counts[player]['Trophy'] += 1

# Open and read the existing leaderboard files to get the old ranks
old_fish_rankings = {}
old_trophy_rankings = {}

with open('leaderboardfish.md', 'r', encoding='utf-8') as file:
    next(file) # Skip first 4 lines
    next(file)
    next(file)
    next(file)
    for line in file:
        if line.startswith("|"):
            parts = line.split("|")
            rank = parts[1].strip().split()[0]
            player = parts[2].strip()
            if '*' in player:
                player = player.rstrip('*')
            player = renamed_chatters.get(player, player)
            while player in renamed_chatters:
                player = renamed_chatters[player]
            old_fish_rankings[player] = int(rank)

with open('leaderboardtrophies.md', 'r', encoding='utf-8') as file:
    next(file) # Skip first 4 lines
    next(file)
    next(file)
    next(file)
    for line in file:
        if line.startswith("|"):
            parts = line.split("|")
            rank = parts[1].strip().split()[0]
            player = parts[2].strip()
            player = renamed_chatters.get(player, player)
            while player in renamed_chatters:
                player = renamed_chatters[player]
            old_trophy_rankings[player] = int(rank)

# Sort players by maximum fish caught and assign ranks with ties handled
rank = 0
prev_max_fish = float('inf')  # Initialize with infinity

# Write the results into a Markdown table for fish leaderboard
with open('leaderboardfish.md', 'w', encoding='utf-8') as file:
    file.write("### Leaderboard for the most fish caught in a single week in tournaments\n\n")
    file.write("| Rank | Player | Fish Caught 🪣 |\n")
    file.write("|------|--------|---------------|\n")
    for player, info in sorted(max_fish_in_week.items(), key=lambda x: x[1]['fish_count'], reverse=True):
        max_fish = info['fish_count']
        bot = info['bot']
        if max_fish >= 20:
            if max_fish < prev_max_fish:
                rank += 1
            # Ranking change
            movement = {}
            old_rank = old_fish_rankings.get(player)
            if old_rank:
                if rank < old_rank:
                    movement[player] = '⬆'
                elif rank > old_rank:
                    movement[player] = '⬇'
                else:
                    movement[player] = ''
            else:
                movement[player] = '🆕'
            if player not in verified_players and bot == 'supibot':
                file.write(f"| {rank} {movement[player]}| {player}* | {max_fish} |\n")
            else:
                file.write(f"| {rank} {movement[player]}| {player} | {max_fish} |\n")
            prev_max_fish = max_fish
    file.write("\n_* = The fish were caught on supibot and the player did not migrate their data over to gofishgame. Because of that their data was not individually verified to be accurate._\n")

# Calculate total points for each player
total_points = defaultdict(float)
for player, counts in player_counts.items():
    total_points[player] = counts['Trophy'] * point_values['Trophy'] + counts['Silver'] * point_values['Silver'] + counts['Bronze'] * point_values['Bronze']

# Sort players by total points and assign positions
sorted_players = sorted(total_points.items(), key=lambda x: x[1], reverse=True)

# Group players with the same points
grouped_players = [(points, list(group)) for points, group in groupby(sorted_players, key=lambda x: x[1])]

# Write the results into a Markdown table for trophies leaderboard
with open('leaderboardtrophies.md', 'w', encoding='utf-8') as file:
    file.write("### Leaderboard for the weekly tournaments\n\n")
    file.write("| Rank | Player | Trophies 🏆 | Silver Medals 🥈 | Bronze Medals 🥉 | Points |\n")
    file.write("|----------|--------|------------|-----------------|-----------------|--------|\n")
    for rank, (points, group) in enumerate(grouped_players, start=1):
        for player, _ in group:
            # Ranking change
            movement = {}
            old_rank = old_trophy_rankings.get(player)
            if old_rank:
                if rank < old_rank:
                    movement[player] = '⬆'
                elif rank > old_rank:
                    movement[player] = '⬇'
                else:
                    movement[player] = ''
            else:
                movement[player] = '🆕'
            file.write(f"| {rank} {movement[player]}| {player} | {player_counts[player]['Trophy']} | {player_counts[player]['Silver']} | {player_counts[player]['Bronze']} | {points} |\n")

File: leaderboards/tournaments.go
package leaderboards

import (
	"bufio"
	"fmt"
	"gofish/playerdata"
	"gofish/utils"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

type PlayerInfo struct {
	FishCount int
	Bot       string
}

type PlayerCounts struct {
	Trophy int
	Silver int
	Bronze int
}

func RunTournaments(chatNames, leaderboard string) {

	// Get the current working directory
	wd, err := os.Getwd()
	if err != nil {
		fmt.Println("Error getting current working directory:", err)
		os.Exit(1)
	}

	// Construct the absolute path to the config file
	configFilePath := filepath.Join(wd, "config.json")

	// Load the config from the constructed file path
	config := utils.LoadConfig(configFilePath)

	// Define point values
	pointValues := map[string]float64{"Trophy": 3, "Silver": 1, "Bronze": 0.5}

	switch chatNames {
	case "all":
		// Process all chats
		for chatName, chat := range config.Chat {
			if !chat.CheckEnabled {
				fmt.Printf("Skipping chat '%s' because check_enabled is false.\n", chatName)
				continue // Skip processing if check_enabled is false
			}

			fmt.Printf("Checking chat '%s'.\n", chatName)
			processTournaments(chatName, chat, pointValues, leaderboard)
		}
	case "":
		fmt.Println("Please specify chat names.")
	default:
		// Process specified chat names
		specifiedchatNames := strings.Split(chatNames, ",")
		for _, chatName := range specifiedchatNames {
			chat, ok := config.Chat[chatName]
			if !ok {
				fmt.Printf("Chat '%s' not found in config.\n", chatName)
				continue
			}
			if !chat.CheckEnabled {
				fmt.Printf("Skipping chat '%s' because check_enabled is false.\n", chatName)
				continue // Skip processing if check_enabled is false
			}

			fmt.Printf("Checking chat '%s'.\n", chatName)
			processTournaments(chatName, chat, pointValues, leaderboard)
		}
	}
}

func processTournaments(chatName string, chat utils.ChatInfo, pointValues map[string]float64, leaderboard string) {

	renamedChatters := playerdata.ReadRenamedChatters()
	cheaters := playerdata.ReadCheaters()

	// Get the current working directory
	wd, err := os.Getwd()
	if err != nil {
		fmt.Println("Error getting current working directory:", err)
		os.Exit(1)
	}
	// Open and read the logs file for the chat
	logsFilePath := filepath.Join(wd, "data", chat.Logs)
	logs, err := os.Open(logsFilePath)
	if err != nil {
		panic(err)
	}
	defer logs.Close()

	fishweekLimit := chat.Fishweeklimit
	if fishweekLimit == 0 {
		fishweekLimit = 20 // Set the default fishweek limit if not specified
	}

	// Define a map to store the maximum fish caught in a week for each player and the bot name
	maxFishInWeek := make(map[string]PlayerInfo)

	// Define a map to store the counts for each player's trophies and medals
	playerCounts := make(map[string]PlayerCounts)

	scanner := bufio.NewScanner(logs)
	for scanner.Scan() {
		line := scanner.Text()

		// Find the player
		playerMatch := regexp.MustCompile(`[@👥]\s?(\w+)`).FindStringSubmatch(line)
		if len(playerMatch) > 0 {
			player := playerMatch[1]

			// Change to the latest name
			newPlayer := renamedChatters[player]
			for newPlayer != "" {
				player = newPlayer
				newPlayer = renamedChatters[player]
			}

			if utils.Contains(cheaters, player) {
				continue // Skip processing for ignored players
			}

			// Get the amount of fish the player caught
			fishMatch := regexp.MustCompile(`(\d+) fish: (\w+)`).FindStringSubmatch(line)
			if len(fishMatch) > 0 {
				fishCount, _ := strconv.Atoi(fishMatch[1])
				botMatch := regexp.MustCompile(`#\w+ \s?(\w+):`).FindStringSubmatch(line)
				if len(botMatch) > 0 {
					bot := botMatch[1]

					// Update the record if the current fish count is greater
					if fishCount > maxFishInWeek[player].FishCount && fishCount >= fishweekLimit {
						maxFishInWeek[player] = PlayerInfo{FishCount: fishCount, Bot: bot}
					}
				}
			}

			// Find all their medals and trophies
			achievements := regexp.MustCompile(`(Victory|champion|runner-up|third)`).FindAllString(line, -1)
			for _, achievement := range achievements {
				switch achievement {
				case "Victory", "champion":
					// Get the player's counts
					counts := playerCounts[player]

					// Update the Trophy count
					counts.Trophy++

					// Assign the updated counts back to the map
					playerCounts[player] = counts
				case "runner-up":
					// Get the player's counts
					counts := playerCounts[player]

					// Update the Silver count
					counts.Silver++

					// Assign the updated counts back to the map
					playerCounts[player] = counts
				case "third":
					// Get the player's counts
					counts := playerCounts[player]

					// Update the Bronze count
					counts.Bronze++

					// Assign the updated counts back to the map
					playerCounts[player] = counts
				}
			}
		}
	}

	// Titles for the leaderboards
	titlefishw := fmt.Sprintf("### Most fish caught in a single week in tournaments in %s's chat\n", chatName)
	titletrophies := fmt.Sprintf("### Leaderboard for the weekly tournaments in %s's chat\n", chatName)

	// Update only the specified leaderboard if the leaderboard flag is provided
	switch leaderboard {
	case "trophy":
		// Write the leaderboard for the weekly tournaments to the file specified in the config
		fmt.Printf("Updating trophies leaderboard for chat '%s'...\n", chatName)
		err = writeTrophiesLeaderboard(chat.Trophies, playerCounts, pointValues, titletrophies)
		if err != nil {
			fmt.Println("Error writing trophies leaderboard:", err)
		} else {
			fmt.Println("Trophies leaderboard updated successfully.")
		}
	case "fishw":
		// Write the leaderboard for the most fish caught in a single week in tournaments to the file specified in the config
		fmt.Printf("Updating fishweek leaderboard for chat '%s' with fish count threshold %d...\n", chatName, fishweekLimit)
		err = writeFishWeekLeaderboard(chat.Fishweek, maxFishInWeek, titlefishw)
		if err != nil {
			fmt.Println("Error writing fishweek leaderboard:", err)
		} else {
			fmt.Println("Fishweek leaderboard updated successfully.")
		}
	default:
		// If the leaderboard flag is not provided, update both leaderboards
		fmt.Printf("Updating trophies leaderboard for chat '%s'...\n", chatName)
		err = writeTrophiesLeaderboard(chat.Trophies, playerCounts, pointValues, titletrophies)
		if err != nil {
			fmt.Println("Error writing trophies leaderboard:", err)
		} else {
			fmt.Println("Trophies leaderboard updated successfully.")
		}

		fmt.Printf("Updating fishweek leaderboard for chat '%s' with fish count threshold %d...\n", chatName, fishweekLimit)
		err = writeFishWeekLeaderboard(chat.Fishweek, maxFishInWeek, titlefishw)
		if err != nil {
			fmt.Println("Error writing fishweek leaderboard:", err)
		} else {
			fmt.Println("Fishweek leaderboard updated successfully.")
		}
	}
}

// Function to write the Trophies leaderboard with emojis indicating ranking change and the change of trophies and medals in brackets
func writeTrophiesLeaderboard(filePath string, playerCounts map[string]PlayerCounts, pointValues map[string]float64, titletrophies string) error {
	// Call ReadOldTrophyRankings to get the trophy rankings
	oldLeaderboardTrophy, err := ReadOldTrophyRankings(filePath)
	if err != nil {
		return err
	}

	// Ensure that the directory exists before attempting to create the file
	err = os.MkdirAll(filepath.Dir(filePath), 0755)
	if err != nil {
		return err
	}

	// Open the file for writing (or create it if it doesn't exist)
	file, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer file.Close()

	// Write the title
	_, err = fmt.Fprintf(file, "%s", titletrophies)
	if err != nil {
		return err
	}

	// Write the header
	_, err = fmt.Fprintln(file, "| Rank | Player | Trophies 🏆 | Silver Medals 🥈 | Bronze Medals 🥉 | Points |")
	_, err = fmt.Fprintln(file, "|------|--------|-------------|------------------|------------------|--------|")
	if err != nil {
		return err
	}

	// Calculate total points for each player
	totalPoints := make(map[string]float64)
	for player, counts := range playerCounts {
		totalPoints[player] = float64(counts.Trophy)*pointValues["Trophy"] + float64(counts.Silver)*pointValues["Silver"] + float64(counts.Bronze)*pointValues["Bronze"]
	}

	// Sort players by total points
	sortedPlayers := utils.SortMapByValueDesc(totalPoints)

	// Write the leaderboard data
	rank := 1
	prevRank := 1
	prevPoints := -1.0
	occupiedRanks := make(map[int]int)

	for _, player := range sortedPlayers {
		points := totalPoints[player]

		// Increment rank only if the count has changed
		if points != prevPoints {
			rank += occupiedRanks[rank] // Increment rank by the number of occupied ranks
			// Reset the count of occupied ranks when count changes
			occupiedRanks[rank] = 1
		} else {
			// Set the rank to the previous rank if the count hasn't changed
			rank = prevRank
			occupiedRanks[rank]++ // Increment the count of occupied ranks
		}

		// Declare found variable in the outer scope
		var found bool

		// Getting the old rank
		oldRank := -1 // Default value if the old rank is not found
		if info, ok := oldLeaderboardTrophy[player]; ok {
			found = true
			oldRank = info.Rank
		}

		changeEmoji := ChangeEmoji(rank, oldRank, found)

		// Compare new counts to old counts and display the difference
		trophiesDifference := playerCounts[player].Trophy - oldLeaderboardTrophy[player].Trophy
		silverDifference := playerCounts[player].Silver - oldLeaderboardTrophy[player].Silver
		bronzeDifference := playerCounts[player].Bronze - oldLeaderboardTrophy[player].Bronze

		// Construct the string with the difference in brackets
		trophyCount := fmt.Sprintf("%d", playerCounts[player].Trophy)
		if trophiesDifference > 0 {
			trophyCount += fmt.Sprintf(" (+%d)", trophiesDifference)
		}

		silverCount := fmt.Sprintf("%d", playerCounts[player].Silver)
		if silverDifference > 0 {
			silverCount += fmt.Sprintf(" (+%d)", silverDifference)
		}

		bronzeCount := fmt.Sprintf("%d", playerCounts[player].Bronze)
		if bronzeDifference > 0 {
			bronzeCount += fmt.Sprintf(" (+%d)", bronzeDifference)
		}

		ranks := Ranks(rank)

		// Write the leaderboard row
		_, err = fmt.Fprintf(file, "| %s %s| %s | %s | %s | %s | %.1f |\n", ranks, changeEmoji, player, trophyCount, silverCount, bronzeCount, totalPoints[player])
		if err != nil {
			return err
		}

		prevPoints = points // Update previous points
		prevRank = rank     // Update previous rank
	}

	return nil
}

// Function to write the Fish Week leaderboard with emojis indicating ranking change
func writeFishWeekLeaderboard(filePath string, maxFishInWeek map[string]PlayerInfo, titlefishw string) error {
	// Call ReadOldFishRankings to get the old fish rankings
	oldLeaderboardFishW, err := ReadOldFishRankings(filePath)
	if err != nil {
		return err
	}

	// Ensure that the directory exists before attempting to create the file
	err = os.MkdirAll(filepath.Dir(filePath), 0755)
	if err != nil {
		return err
	}

	// Open the file for writing (or create it if it doesn't exist)
	file, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer file.Close()

	// Write the title
	_, err = fmt.Fprintf(file, "%s", titlefishw)
	if err != nil {
		return err
	}

	// Write the header
	_, err = fmt.Fprintln(file, "| Rank | Player | Fish Caught 🪣 |")
	_, err = fmt.Fprintln(file, "|------|--------|---------------|")
	if err != nil {
		return err
	}

	verifiedPlayers := playerdata.ReadVerifiedPlayers()

	// Calculate total fish caught for each player
	totalFishCaught := make(map[string]int)
	for player, info := range maxFishInWeek {
		totalFishCaught[player] = info.FishCount
	}

	// Sort players by total fish caught
	sortedPlayers := utils.SortMapByValueDescInt(totalFishCaught)

	// Write the leaderboard data
	rank := 1
	prevRank := 1
	prevFishCount := -1
	occupiedRanks := make(map[int]int)

	for _, player := range sortedPlayers {
		fishCount := totalFishCaught[player]

		// Increment rank only if the count has changed
		if fishCount != prevFishCount {
			rank += occupiedRanks[rank] // Increment rank by the number of occupied ranks
			// Reset the count of occupied ranks when count changes
			occupiedRanks[rank] = 1
		} else {
			// Set the rank to the previous rank if the count hasn't changed
			rank = prevRank
			occupiedRanks[rank]++ // Increment the count of occupied ranks
		}

		// Declare found variable in the outer scope
		var found bool

		// Getting the old rank
		oldRank := -1 // Default value if the old rank is not found
		if info, ok := oldLeaderboardFishW[player]; ok {
			found = true
			oldRank = info.Rank
		}

		changeEmoji := ChangeEmoji(rank, oldRank, found)

		// Construct the string with the difference in brackets
		fishweekDifference := fishCount - oldLeaderboardFishW[player].Count
		fishWeekCount := fmt.Sprintf("%d", fishCount)
		if fishweekDifference > 0 && oldLeaderboardFishW[player].Count > 0 {
			fishWeekCount += fmt.Sprintf(" (+%d)", fishweekDifference)
		}

		// Write the leaderboard row with change indication
		botIndicator := ""
		if maxFishInWeek[player].Bot == "supibot" && !utils.Contains(verifiedPlayers, player) {
			botIndicator = "*"
		}

		ranks := Ranks(rank)

		_, err = fmt.Fprintf(file, "| %s %s| %s%s | %s |\n", ranks, changeEmoji, player, botIndicator, fishWeekCount)
		if err != nil {
			return err
		}

		prevFishCount = fishCount // Update previous fish count
		prevRank = rank           // Update previous rank

	}

	// Write the note
	_, err = fmt.Fprintln(file, "\n_* = The fish were caught on supibot and the player did not migrate their data over to gofishgame. Because of that their data was not individually verified to be accurate._")
	if err != nil {
		return err
	}

	return nil
}
